### ISSUE TITLE
ConvNd function leaks memory

### ISSUE THREAD

**Issue by apaszke** [pytorch/pytorch#3835]
Originally reported by @dmarnerides in #3743:

---

I can reproduce the memory leak with this:

```
import gc
import resource
import torch
from torch import nn, autograd
from torch.autograd import Variable

class Network(nn.Module):
    def __init__(self):
        super(Network, self).__init__()
        self.main = nn.Sequential(
            nn.Conv2d(1, 1, 1, 1),
            nn.Conv2d(1, 1, 1, 1),
        )
    def forward(self, v_x):
        return self.main(v_x).view(v_x.size(0), 1)

net = Network()

i = 0
while True:
    v_in = Variable(torch.Tensor(2,1,1,1), requires_grad=True)
    grad_out = Variable(torch.ones(2,1,1,1))

    gradient = autograd.grad(outputs=net(v_in), inputs=v_in,
                             grad_outputs=grad_out,
                             create_graph=True, retain_graph=True, 
                             only_inputs=True)[0]
    gradient.mean().backward()

    i += 1
    if i % 512 == 0:
        gc.collect()
        max_mem_used = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
        print("{:.2f} MB".format(max_mem_used / 1024))
```

However the leak disappears when only one convolution is used:
```
class Network(nn.Module):
    def __init__(self):
        super(Network, self).__init__()
        self.main = nn.Sequential(
            nn.Conv2d(1, 1, 1, 1),
        )
    def forward(self, v_x):
        return self.main(v_x).view(v_x.size(0), 1)
```

Also, if nn.Linear is used, there is no leak.

Torch version: 0.4.0a0+8ebf18b

**Comment by soumith** [comment]
as I am going through issues, I want to add context that @colesbury identified the issue. It's a reference cycle. Solutions TBD

**Comment by dmarnerides** [comment]
Hey, thanks for working on this. Will these fixes be included in the 0.3 binaries or is this branch already far ahead?

**Comment by li-roy** [comment]
Closed by #3984 and #3970
